### PR TITLE
add taskStats support

### PR DESCRIPTION
**Pull request by matt-deboer**
Adds support for `?embed=apps.taskStats` optional parameter

### CODE REVIEW

File: application.go [hunk 1]
```diff
@@ -56,41 +56,42 @@ type Port struct {
...
+	BackoffSeconds        *float64               `json:"backoffSeconds,omitempty"`
+	BackoffFactor         *float64               `json:"backoffFactor,omitempty"`
+	MaxLaunchDelaySeconds *float64               `json:"maxLaunchDelaySeconds,omitempty"`
+	Deployments           []map[string]string    `json:"deployments,omitempty"`
+	Dependencies          []string               `json:"dependencies"`
+	TasksRunning          int                    `json:"tasksRunning,omitempty"`
+	TasksStaged           int                    `json:"tasksStaged,omitempty"`
+	TasksHealthy          int                    `json:"tasksHealthy,omitempty"`
+	TasksUnhealthy        int                    `json:"tasksUnhealthy,omitempty"`
+	TaskStats             *map[string]*TaskStats `json:"taskStats,omitempty"`
```
Comment (timoreimann) [inline application.go]: I think we should be able to remove all pointers as this is a read-only parameter.

File: application.go [hunk 2]
```diff
@@ -130,6 +131,17 @@ type DeleteAppOpts struct {
 	Force bool `url:"force,omitempty"`
 }
 
+// TaskStats is a container for Stats
+type TaskStats struct {
+	Stats *Stats `json:"stats,omitempty"`
+}
+
+// Stats is a collection of aggregate statistics about an application's tasks
+type Stats struct {
```
Comment (timoreimann) [inline application.go]: Depointerize both struct fields here too.

File: tests/rest-api/methods.yml [hunk 3]
```diff
@@ -203,6 +203,227 @@
         }
     ]
     }
+- uri: /v2/apps?embed=apps.taskStats
+  method: GET
+  content: |
+    {
+    "apps": [
+        {
+            "args": null,
```
Comment (timoreimann) [inline tests/rest-api/methods.yml]: Could you please reduce the test data definition to just the fields we actually need for testing?

Comment (matt-deboer) [inline tests/rest-api/methods.yml]: tried to reduce further than this, but general app deserialization fails at that point.

**Comment by matt-deboer** [comment]
ok; done...look good?

File: tests/rest-api/methods.yml [hunk 4]
```diff
@@ -203,6 +203,102 @@
...
+            "mem": 64.0,
+            "ports": [],
+            "requirePorts": false,
+            "storeUrls": [],
+            "tasksRunning": 2,
+            "tasksStaged": 0,
+            "taskStats": {
+                "startedAfterLastScaling": {
+                    "stats": {
+                    "counts": {
```
Comment (timoreimann) [inline tests/rest-api/methods.yml]: This block needs to be indented one more level.

File: tests/rest-api/methods.yml [hunk 5]
```diff
@@ -203,6 +203,102 @@
...
+                    },
+                    "lifeTime": {
+                        "averageSeconds": 17024.575,
+                        "medianSeconds": 17024.575
+                    }
+                    }
+                },
+                "withLatestConfig": {
+                    "stats": {
+                    "counts": {
```
Comment (timoreimann) [inline tests/rest-api/methods.yml]: Indent here too.

File: tests/rest-api/methods.yml [hunk 6]
```diff
@@ -203,6 +203,102 @@
...
+                    },
+                    "lifeTime": {
+                        "averageSeconds": 17024.575,
+                        "medianSeconds": 17024.575
+                    }
+                    }
+                },
+                "totalSummary": {
+                    "stats": {
+                    "counts": {
```
Comment (timoreimann) [inline tests/rest-api/methods.yml]: Indent.

File: tests/rest-api/methods.yml [hunk 7]
```diff
@@ -203,6 +203,102 @@
...
+                },
+                "withLatestConfig": {
+                    "stats": {
+                    "counts": {
+                        "staged": 0,
+                        "running": 1,
+                        "healthy": 1,
+                        "unhealthy": 0
+                    },
+                    "lifeTime": {
```
Comment (timoreimann) [inline tests/rest-api/methods.yml]: Indent here too.

File: tests/rest-api/methods.yml [hunk 8]
```diff
@@ -203,6 +203,102 @@
...
+                },
+                "totalSummary": {
+                    "stats": {
+                    "counts": {
+                        "staged": 0,
+                        "running": 1,
+                        "healthy": 1,
+                        "unhealthy": 0
+                    },
+                    "lifeTime": {
```
Comment (timoreimann) [inline tests/rest-api/methods.yml]: Indent here too.

File: tests/rest-api/methods.yml (hunk 3 above)
Comment (timoreimann) [inline tests/rest-api/methods.yml]: That should only be the case when you are producing incorrect JSON. (Which can easily happen when you modify `methods.yml`; at least, it keeps happening to me.)

I could successfully run your tests with the following, minimalistic JSON response:

```json
{
    "apps": [
        {
            "taskStats": {
                "startedAfterLastScaling": {
                    "stats": {
                      "counts": {
                          "staged": 0,
                          "running": 1,
                          "healthy": 1,
                          "unhealthy": 0
                      },
                      "lifeTime": {
                          "averageSeconds": 17024.575,
                          "medianSeconds": 17024.575
                      }
                    }
                }
            }
        }
    ]
    }
```

File: application.go [hunk 9]
```diff
@@ -130,6 +131,17 @@ type DeleteAppOpts struct {
...
 
+// TaskStats is a container for Stats
+type TaskStats struct {
+	Stats Stats `json:"stats"`
+}
+
+// Stats is a collection of aggregate statistics about an application's tasks
+type Stats struct {
+	Counts   map[string]int     `json:"counts"`
+	LifeTime map[string]float64 `json:"lifetime"`
```
Comment (timoreimann) [inline application.go]: I'm not a native English speaker but shouldn't this be written `Lifetime`?

File: tests/rest-api/methods.yml (hunk 3 above)
Comment (timoreimann) [inline tests/rest-api/methods.yml]: That should only be the case when you are producing incorrect JSON. (Which can easily happen when you modify `methods.yml`; at least, it keeps happening to me.) JSON deserialization only cares about JSON structure, missing fields should just be skipped.

I could successfully run your tests with the following, minimalistic JSON response:

```json
{
    "apps": [
        {
            "taskStats": {
                "startedAfterLastScaling": {
                    "stats": {
                      "counts": {
                          "staged": 0,
                          "running": 1,
                          "healthy": 1,
                          "unhealthy": 0
                      },
                      "lifeTime": {
                          "averageSeconds": 17024.575,
                          "medianSeconds": 17024.575
                      }
                    }
                }
            }
        }
    ]
    }
```

Comment (matt-deboer) [inline tests/rest-api/methods.yml]: you're right; thanks. not sure why that was failing for me ealier...

File: application.go (hunk 9 above)
Comment (matt-deboer) [inline application.go]: yes, I would normally write it that way; though it seems that `lifeTime` is the json produced by the marathon API...happy to change the go field name though

Comment (timoreimann) [inline application.go]: If the JSON field name uses a capital T, then I agree we should have it in the variable name as well. You did not use a capital T in the JSON tag name though -- is that a typo?

Whatever the right spelling is, we should keep it in sync.

File: application_test.go [hunk 10]
```diff
@@ -297,6 +297,20 @@ func TestApplications(t *testing.T) {
...
+
+	v := url.Values{}
+	v.Set("embed", "apps.taskStats")
+	applications, err := endpoint.Client.Applications(v)
+	assert.NoError(t, err)
+	assert.NotNil(t, applications)
+	assert.Equal(t, len(applications.Apps), 1)
+	assert.NotNil(t, applications.Apps[0].TaskStats)
+	assert.Equal(t, applications.Apps[0].TaskStats["startedAfterLastScaling"].Stats.Counts["healthy"], 1)
+}
```
Comment (timoreimann) [inline application_test.go]: Seems like this comment got lost somehow:

Can we add an assertion for TaskStats.LifeTime too please?

**Comment by timoreimann** [comment]
LGTM -- thanks for bearing with me. :-)

**Pull request merged**